### PR TITLE
add gcp to stores

### PR DESCRIPTION
**Pull request by arianxplore**
@mahdi13  Mahdi, Please Review this PR and let me know if it's Looks good to you.
and let me know if i can PR to the base Repository too.
Regards 
AE

### CODE REVIEW

File: sqlalchemy_media/stores/gcp.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+from .base import Store
+from ..typing_ import FileLike
+
+
+class GCPStore(Store):
+    """
+    Google Cloud Storage Implements Store base.
+    """
+
+    def __init__(self, bucket: str, service_account_json: str, acl: str = 'private'):
```
Comment (mahdi13) [inline sqlalchemy_media/stores/gcp.py]: @ribrea Please add documentation for the parameters

File: sqlalchemy_media/optionals.py [hunk 2]
```diff
@@ -70,3 +68,22 @@ def ensure_paramiko():
 
     if paramiko is None:  # pragma: no cover
         raise OptionalPackageRequirementError('paramiko')
+
+
+# paramiko
```
Comment (mahdi13) [inline sqlalchemy_media/optionals.py]: @ribrea This is not correct. It's related the previous dependency

File: sqlalchemy_media/stores/gcp.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+from google.cloud import storage
+from google.cloud.exceptions import NotFound
+
+from sqlalchemy_media.exceptions import GCPError
+from sqlalchemy_media.optionals import ensure_gcs
+from .base import Store
+from ..typing_ import FileLike
+
+
+class GCPStore(Store):
```
Comment (mahdi13) [inline sqlalchemy_media/stores/gcp.py]: @ribrea GCP Means Google Cloud Platform. I think  the more accurate term is GoogleCloudStorge (apparently GCS is not a valid term too)

**Review by mahdi13 — CHANGES_REQUESTED** [review]

File: sqlalchemy_media/stores/gcp.py (hunk 3 above)
Comment (arianxplore) [reply sqlalchemy_media/stores/gcp.py]: that's alright i just took the name from task title.
Fixes #1f98f21

File: sqlalchemy_media/optionals.py (hunk 2 above)
Comment (arianxplore) [reply sqlalchemy_media/optionals.py]: Sorry, i just copy the Paramiko ensure but forgot to change the comment.
Fixes #73bb6d3

File: sqlalchemy_media/stores/gcp.py (hunk 1 above)
Comment (arianxplore) [reply sqlalchemy_media/stores/gcp.py]: I tough it's clear enough, i add param document.
Fixed #4d8dde9

**Review by mahdi13 — APPROVED** [review]

**Comment by mahdi13** [comment]
@ribrea Thanks, LGTM, please proceed with the upstream [pylover](https://github.com/pylover/sqlalchemy-media) PR

File: sqlalchemy_media/stores/GoogleCloudStorge.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+from google.cloud import storage
+from google.cloud.exceptions import NotFound
+
+from sqlalchemy_media.exceptions import GCPError
+from sqlalchemy_media.optionals import ensure_gcs
+from .base import Store
+from ..typing_ import FileLike
+
+
+class GoogleCloudStorge(Store):
```
Comment (mahdi13) [inline sqlalchemy_media/stores/GoogleCloudStorge.py:15]: `GoogleCloudStorge` -> `GoogleCloudStorage`
@ribrea

File: sqlalchemy_media/stores/GoogleCloudStorge.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    def open(self, filename: str, mode: str = 'rb') -> FileLike:
+        """
+        Download a file as byte
+        """
+        ensure_gcs()
+        try:
+            bucket = self._get_or_create_bucket()
+            blob = bucket.blob(filename)
+            file_byte = blob.download_to_filename()
```
Comment (mahdi13) [inline sqlalchemy_media/stores/GoogleCloudStorge.py:76]: `file_byte = blob.download_as_bytes()`
@ribrea

File: sqlalchemy_media/stores/GoogleCloudStorge.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        try:
+            bucket = self._get_or_create_bucket()
+            blob = bucket.blob(file_name)
+            blob.upload_from_string(data)
+            if self.acl == 'public':
+                blob.make_public()
+        except Exception as e:
+            raise GCPError(e)
+
+    def locate(self, attachment) -> str:
```
Comment (mahdi13) [inline sqlalchemy_media/stores/GoogleCloudStorge.py:57]: @ribrea Most of the time the resource is protected, so there won't be any public URL. I suggest using the same method as s3 store: 
```
    def locate(self, attachment) -> str:
        if self.cdn_url:
            base_url = self.cdn_url
        else:
            base_url = self.base_url
        return '%s/%s' % (base_url, attachment.path)
```

Comment (arianxplore) [reply sqlalchemy_media/stores/GoogleCloudStorge.py:57]: Google's Doing the same ting

````
@property
    def public_url(self):
        """The public URL for this blob.

        Use :meth:`make_public` to enable anonymous access via the returned
        URL.

        :rtype: `string`
        :returns: The public URL for this blob.
        """
        return "{storage_base_url}/{bucket_name}/{quoted_name}".format(
            storage_base_url=_API_ACCESS_ENDPOINT,
            bucket_name=self.bucket.name,
            quoted_name=_quote(self.name, safe=b"/~"),
        )

````

**Pull request merged**